synapse/http/server.py

```diff
@@ -499,6 +500,71 @@ class RootOptionsRedirectResource(OptionsResource, RootRedirect):
     pass
 
 
+@implementer(interfaces.IPullProducer)
+class _ByteProducer:
+    """
+    Iteratively write bytes to the request.
+    """
+
+    # The minimum number of bytes for each chunk. Note that the last chunk will
+    # usually be smaller than this.
+    min_chunk_size = 1024
+
+    def __init__(
+        self, request: Request, iterator: Iterator[bytes],
+    ):
+        self._request = request
+        self._iterator = iterator
+
+    def start(self):
+        self._request.registerProducer(self, False)
+
+    def _send_data(self, data: List[bytes]) -> None:
+        """
+        Send a list of strings as a response to the request.
+        """
+        if not data:
+            return
+        self._request.write(b"".join(data))
+
+    def resumeProducing(self):
+        # We've stopped producing in the meantime.
+        if not self._request:
+            return
+
+        # Get the next chunk and write it to the request. Calling write will
+        # spin the reactor (and might be re-entrant).
+        buffer = []
+        buffered_bytes = 0
+        while buffered_bytes < self.min_chunk_size:
+            try:
+                data = next(self._iterator)
+                buffer.append(data)
+                buffered_bytes += len(data)
+            except StopIteration:
+                # Everything is serialized, write any data, then finalize the
+                # producer.
+                self._send_data(buffer)
+                self._request.unregisterProducer()
+                self._request.finish()
+                self.stopProducing()
+                return
+
+        self._send_data(buffer)
+
+    def stopProducing(self):
+        self._generator = None  # type: ignore
```

Comment:
What is this?

Reply:
Are you asking why are we setting it to none, or why is the type ignored?

It is set to none because that's what the Twisted code that I cribbed from did. 😄 I believe this is to ensure that we ensure any cyclic references are broken. This isn't necessary in our implementation though.

The type being ignored is because `_generator` isn't set to `Optional[...]` earlier.

Reply:
But it's not used or set anywhere else? Or does something come along and set it on this object?

Reply:
Ah...`self._iterator` used to be called `self._generator` 😄 I suspect I should just remove this.

Reply:
I've removed this.